### PR TITLE
Gp practice search

### PR DESCRIPTION
**Pull request by MarkC33**
Generic general search service added and implementation for GpPracticeSearch added.

Includes tasks from user stories #9915 and #9957

commits are also included from CodeSmellReduction branch

### CODE REVIEW

File: Mep.Business/Services/GpPracticeSearchService.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  public class GpPracticeSearchService : GeneralSearchServiceBase, IModelGeneralSearchService<GpPractice>
+  {
+    public GpPracticeSearchService(ApplicationContext context, IMapper mapper)
+      : base("GpPractice", context, mapper)
+    {
+    }
+
+    public override async Task<IEnumerable<GeneralSearchResult>> SearchAsync(string searchString)
+    {
+      if (searchString != "")
```
Comment (neildavies11) [inline Mep.Business/Services/GpPracticeSearchService.cs]: A better check for != "" is to use string.IsNullOrWhiteSpace(searchString)
Also even in cases like this when there isn't a lot of code for the true block let's always have it first so the sequence is always IF {TRUE BLOCK} ELSE {FALSE BLOCK}. Keeping this consistant makes all code easier to understand quickly.

File: Mep.Business/Services/GpPracticeSearchService.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      {
+        IEnumerable<GeneralSearchResult> entities =
+          await _context.GpPractices
+          .WhereIsActiveOrActiveOnly(true)
```
Comment (neildavies11) [inline Mep.Business/Services/GpPracticeSearchService.cs]: Like the use of this rather than a stand alone Where(gp => gp.IsActive) clause :)

**Comment by MarkC33** [comment]
> Let's change all the Int64's to longs.

Int64 instances change to long

File: Mep.Api/ViewModels/Profiles/ReferralPatientSearchProfile.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+using System.Linq;
+
+namespace Mep.Api.ViewModels.Profiles
+{
+    public class ReferralPatientSearchProfile : Profile
+    {
+        public ReferralPatientSearchProfile()
+        {
+            CreateMap<BusinessModels.Patient, ReferralPatientSearch>()
+            .ForMember(dest => dest.CcgDetails, opt => opt.MapFrom(src => src.Ccg.Name))
```
[the remark this answers is not in the document]
Comment (MarkC33) [reply Mep.Api/ViewModels/Profiles/ReferralPatientSearchProfile.cs]: Changed to CcgName

File: Mep.Api/ViewModels/ReferralPatientSearch.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    public string AlternativeIdentifier { get; set; }
+    public int? CcgId { get; set; }
+    public int? GpPracticeId { get; set; }
+    public Int64? NhsNumber { get; set; }
+    [MaxLength(10)]
+    public string ResidentialPostcode { get; set; }
+    public int? CurrentReferralId { get; set; }
+    public int? PatientId { get; set; }
+    public string GpPracticeDetails { get; set; }
+    public string CcgDetails { get; set; }
```
[the remark this answers is not in the document]
Comment (MarkC33) [reply Mep.Api/ViewModels/ReferralPatientSearch.cs]: Changed to CcgName

File: Mep.Api/ViewModels/Profiles/ReferralPatientSearchProfile.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+            .ForMember(dest => dest.PatientId, opt => opt.MapFrom(src => src.Id))
+            .ForMember(
+              dest => dest.GpPracticeDetails, 
+              opt => opt.MapFrom(src => src.GpPractice.Name + " " + src.GpPractice.Postcode)
```
[the remark this answers is not in the document]
Comment (MarkC33) [reply Mep.Api/ViewModels/Profiles/ReferralPatientSearchProfile.cs]: Updated to a more descriptive name

File: Mep.Api/ViewModels/ReferralPatientSearch.cs [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    [MaxLength(200)]
+    public string AlternativeIdentifier { get; set; }
+    public int? CcgId { get; set; }
+    public int? GpPracticeId { get; set; }
+    public Int64? NhsNumber { get; set; }
+    [MaxLength(10)]
+    public string ResidentialPostcode { get; set; }
+    public int? CurrentReferralId { get; set; }
+    public int? PatientId { get; set; }
+    public string GpPracticeDetails { get; set; }
```
[the remark this answers is not in the document]
Comment (MarkC33) [reply Mep.Api/ViewModels/ReferralPatientSearch.cs]: Updated to a more descriptive name

File: Mep.Api/ViewModels/ReferralPatientSearch.cs [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+using System.ComponentModel.DataAnnotations;
+namespace Mep.Api.ViewModels
+{
+  public class ReferralPatientSearch
+  {
+    [MaxLength(200)]
+    public string AlternativeIdentifier { get; set; }
+    public int? CcgId { get; set; }
+    public int? GpPracticeId { get; set; }
+    public Int64? NhsNumber { get; set; }
```
[the remark this answers is not in the document]
Comment (MarkC33) [reply Mep.Api/ViewModels/ReferralPatientSearch.cs]: Updated

File: Mep.Business/Services/GpPracticeSearchService.cs (hunk 1 above)
Comment (MarkC33) [reply Mep.Business/Services/GpPracticeSearchService.cs]: Improved readability of code

File: Mep.Api/Startup.cs [hunk 8]
```diff
@@ -53,6 +52,17 @@ public void ConfigureServices(IServiceCollection services)
       services.AddScoped<IModelService<Speciality>, SpecialityService>();
       services.AddScoped<IModelService<User>, UserService>();
       services.AddScoped<IModelSimpleSearchService<AvailableDoctor, Business.Models.SearchModels.AvailableDoctorSearch>, AvailableDoctorService>();
+
```
[the remark this answers is not in the document]
Comment (MarkC33) [reply Mep.Api/Startup.cs]: CORS code moved to a new commit

**Pull request merged**